Math/src/main/java/org/tribuo/math/distributions/MultivariateNormalDistribution.java

```diff
@@ -91,62 +99,278 @@ public MultivariateNormalDistribution(DenseVector means, DenseMatrix covariance,
     }
 
     /**
      * Sample a vector from this multivariate normal distribution.
      * @return A sample from this distribution.
      */
+    @Override
     public DenseVector sampleVector() {
+        return sampleVector(rng);
+    }
+
+    /**
+     * Sample a vector from this multivariate normal distribution.
+     * @return A sample from this distribution.
+     */
+    @Override
+    public DenseVector sampleVector(RandomGenerator otherRNG) {
         DenseVector sampled = new DenseVector(means.size());
         for (int i = 0; i < means.size(); i++) {
-            sampled.set(i,rng.nextGaussian());
+            sampled.set(i, otherRNG.nextGaussian());
         }
 
-        sampled = samplingCovariance.leftMultiply(sampled);
+        switch (type) {
+            case FULL -> sampled = samplingCovariance.leftMultiply(sampled);
+            case DIAGONAL -> sampled.hadamardProductInPlace(covarianceVector);
+            case SPHERICAL -> sampled.scaleInPlace(variance);
+        }
+
+        sampled.intersectAndAddInPlace(means);
 
-        return means.add(sampled);
+        return sampled;
     }
 
     /**
-     * Sample a vector from this multivariate normal distribution.
-     * @return A sample from this distribution.
+     * Gets a copy of the mean vector.
+     * @return A copy of the mean vector.
+     */
+    public DenseVector means() {
+        return means.copy();
+    }
+
+    /**
+     * Gets a copy of the covariance, either a {@link DenseMatrix} if it's full rank,
+     * or a {@link DenseVector} if it's diagonal or spherical.
+     * @return The covariance.
+     */
+    public Tensor covariance() {
+        return switch (type) {
+            case FULL -> covarianceMatrix.copy();
+            case DIAGONAL -> covarianceVector.copy();
+            case SPHERICAL -> new DenseVector(means.size(), variance);
+        };
+    }
+
+    /**
+     * Compute the log probability of the input under this multivariate normal distribution.
+     * @param input The input to compute.
+     * @return The log probability.
      */
-    public double[] sampleArray() {
-        return sampleVector().toArray();
+    public double logProbability(SGDVector input) {
+        return logProbability(input, means, precision, determinant, type);
+    }
+
+    public static double logProbability(SGDVector input, DenseVector mean, Tensor precision, double determinant, CovarianceType type) {
+        // p(input|mean, variance) = \frac{1}{(2\pi)^{d/2} determinant^{1/2}} e^{-1/2 * (input - mean)^T * precision * (input - mean)}
+        // log p(i|mu,sigma) = - log ({2\pi}^{d/2}) - log (determinant^{1/2}) + (-1/2 * (i - mu)^T * precision * (i - mu))
+        double scalar = (- (mean.size() / 2.0) * Math.log(2 * Math.PI)) - (Math.log(determinant) / 2.0);
+        DenseVector diff = (DenseVector) input.subtract(mean);
+        double distance = switch (type) {
+            case FULL -> {
+                DenseMatrix precMat = (DenseMatrix) precision;
```

Comment:
I know we've done all the checking already in the construction of the object, but this static method is public and doesn't do any checking. Are we ok with throwing ClassCastExceptions here? It's ok if the answer is yes, but then this method should grow some documentation about being a bit sharp, maybe.

Reply:
Sure, I'll fix that.